### DISCUSSION TITLE
using "instance" field

### DISCUSSION THREAD

**Discussion by maxtoroq** [dotnet/csharplang#1186] [General]
While `using static` is a nice feature, static classes have inherent limitations. I think the same idea can be applied to fields, or even properties. e.g.:

```csharp
class Program {

   using readonly TextWriter log = Console.Out;

   public Program() { }

   public Program(TextWriter log) {
      this.log = log;
   }

   public void Run() {
      WriteLine("Starting");
   }
}
```

I understand this looks a lot like #114, but it's a completely different feature. Unfortunately, the `using` keyword has already too many uses.

Also, if #218 is implemented, then it would be better to not have it on the member declaration, but on it's own `using` declaration, e.g.:

```csharp
class Program {
   using instance log;
   readonly TextWriter log;
```

In the above example, the `instance` keyword is used to specify a member of the current type.

**Comment by CyrusNajmabadi** [comment]
i don't understand how your first example is different from:

```c#
class Program {
   // using -> private
   private readonly TextWriter log = Console.Out;

   public Program() { }

   public Program(TextWriter log) {
      this.log = log;
   }

   public void Run() {
      WriteLine("Starting");
   }
}
```

**Comment by bondsbw** [comment]
@CyrusNajmabadi Looks like it is the `WriteLine`, yours would need to be changed to

```c#
public void Run() {
   log.WriteLine("Starting");
}
```

**Comment by bondsbw** [comment]
This can also be accomplished for individual methods via
```c#
Action<string> WriteLine => log.WriteLine;
```
but the proposal would provide a way to bring everything into scope instead of just a single member.

**Comment by bondsbw** [comment]
If we add access modifiers (e.g. `public using`) then we effectively have an implementation of [member delegation](https://github.com/dotnet/csharplang/issues/477):

```c#
var program = new Program(myLogger);
program.WriteLine("Some string");
```

**Comment by maxtoroq** [comment]
@bondsbw 
> If we add access modifiers (e.g. public using) then we effectively have an implementation of member delegation:

That's not the purpose of the feature, it should work like `using static`, e.g. `WriteLine` is only visible in the (partial) class where the `using` declaration is present.

**Comment by bondsbw** [comment]
Fair enough.  But if member delegation is more popular, influencing its design might get your proposed feature for free.
